client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/ApplicationResourcePrivileges.java

```diff
@@ -67,8 +66,8 @@
     }
 
     private final String application;
-    private final Set<String> privileges;
-    private final Set<String> resources;
+    private final List<String> privileges;
```

Comment:
Why the conversion of these types? the Set.copyOf seemed fine in the ctor before (would still work with changing the ctor args to List). I would assume these can get larger in extreme cases, and would not want to degrade to linear lookups.

Reply:
@rjernst Sure, let me explain the change here. This change here is in the high-level REST client code. Let's take a look at the server code. The server code here is `org.elasticsearch.xpack.core.security.authz.RoleDescriptor$ApplicationResourcePrivileges`. There, we see that `privileges` and `resources` are declared as `String[]`: https://github.com/elastic/elasticsearch/blob/fe36861ada3e75259ff601a9f6b0e1cd5fbb8afe/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java#L840-L841

Further, when these are parsed, and the objects are constructed, no effort is made to de-duplicate any values in these arrays: https://github.com/elastic/elasticsearch/blob/fe36861ada3e75259ff601a9f6b0e1cd5fbb8afe/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java#L835-L836

So server side, duplicates are indeed possible. Therefore, I posit that having the client-side classes do de-duplication is a mistake, that we should be handling these consistently on the server side and on the client side.

Lastly, in most of these tests, we are requiring on the iteration order of these collections (the ones that I changed from a `Set` to `List`). Of course, `Set` does not give us any control over the ordering of the values. This was not a problem when the implementation detail is that we were constructing a `HashSet` and the creating a copy via `Collections.unmodifiableSet(new HashSet<>(source))`; that would indeed maintain iteration order since both collections are `HashSet`s which while they do not have a guaranteed ordering, should at least be consistent given the same elements (although not necessarily, it's not guaranteed at all). This arises as a problem when trying to use `Set#copyOf` because now we are no longer necessarily using the same `Set` implementation. In fact `Set#copyOf` is going to use an `ImmutableSet` which does in fact have a different ordering than a `HashSet` due to, again, implementation details.

So, given that:
 - client and server side should match in implementation, either both allowing duplicates or neither
 - and in some tests we are relying on the iteration order

I elected to use a collection that is consistent with the server side behavior, and also gives us a guaranteed iteration order.

This reasoning applies to all the places in the security code base where I made this change.